pulsar-broker/src/test/java/org/apache/pulsar/compaction/CompactedTopicImplTest.java

```diff
@@ -0,0 +1,147 @@
+        ORIGIN_SPARSE_LEDGER.put(Long.MAX_VALUE, Long.MAX_VALUE);
+    }
+
+    @DataProvider(name = "argsForFindStartPointLoop")
+    public Object[][] argsForFindStartPointLoop() {
+        return new Object[][]{
+                {0, 100, 0},// first value.
+                {0, 100, 1},// second value.
+                {0, 100, 1003},// not first value.
+                {0, 100, 1015},// value not exists.
+                {3, 40, 50},// less than first value & find in a range.
+                {3, 40, 1002},// first value & find in a range.
+                {3, 40, 1003},// second value & find in a range.
+                {3, 40, 1010},// not first value & find in a range.
+                {3, 40, 1015}// value not exists & find in a range.
+        };
+    }
+
+    private static CacheLoader<Long, MessageIdData> mockCacheLoader(long start, long end, final long targetMessageId,
+                                                                    AtomicLong bingoMarker) {
+        // Mock ledger.
+        final TreeMap<Long, Long> sparseLedger = new TreeMap<>();
+        sparseLedger.putAll(ORIGIN_SPARSE_LEDGER.subMap(start, end + 1));
+        sparseLedger.put(Long.MAX_VALUE, Long.MAX_VALUE);
+
+        Function<Long, Long> findMessageIdFunc = entryId -> sparseLedger.ceilingEntry(entryId).getValue();
+
+        // Calculate the correct position.
+        for (long i = start; i <= end; i++) {
+            if (findMessageIdFunc.apply(i) >= targetMessageId) {
+                bingoMarker.set(i);
+                break;
+            }
+        }
+
+        return new CacheLoader<Long, MessageIdData>() {
+            @Override
+            public @Nullable MessageIdData load(@NonNull Long entryId) throws Exception {
+                MessageIdData messageIdData = new MessageIdData();
+                messageIdData.setLedgerId(DEFAULT_LEDGER_ID);
+                messageIdData.setEntryId(findMessageIdFunc.apply(entryId));
+                return messageIdData;
+            }
+        };
+    }
+
+    @Test(dataProvider = "argsForFindStartPointLoop")
+    public void testFindStartPointLoop(long start, long end, long targetMessageId) {
+        AtomicLong bingoMarker = new AtomicLong();
+        // Mock cache.
+        AsyncLoadingCache<Long, MessageIdData> cache = Caffeine.newBuilder()
+                .buildAsync(mockCacheLoader(start, end, targetMessageId, bingoMarker));
+        // Do test.
+        PositionImpl targetPosition = PositionImpl.get(DEFAULT_LEDGER_ID, targetMessageId);
+        CompletableFuture<Long> promise = new CompletableFuture<>();
+        CompactedTopicImpl.findStartPointLoop(targetPosition, start, end, promise, cache);
+        long result = promise.join();
+        assertEquals(result, bingoMarker.get());
+    }
+
+    /**
+     * Why should we check the recursion number of "findStartPointLoop", see: #17976
+     */
+    @Test
+    public void testRecursionNumberOfFindStartPointLoop() throws Exception {
```

Comment:
Seems that this function won't throw any `Exceptions`.

Reply:
Already removed. Thanks